src/main/kotlin/com/openlattice/indexing/BackgroundExpiredDataDeletionService.kt

```diff
@@ -0,0 +1,250 @@
+import com.openlattice.postgres.PostgresTable.DATA
+import com.openlattice.postgres.PostgresTable.IDS
+import com.openlattice.postgres.ResultSetAdapters
+import com.openlattice.postgres.streams.PostgresIterable
+import com.openlattice.postgres.streams.StatementHolder
+import com.zaxxer.hikari.HikariDataSource
+import org.apache.olingo.commons.api.edm.EdmPrimitiveTypeKind
+import org.slf4j.LoggerFactory
+import org.springframework.scheduling.annotation.Scheduled
+import java.sql.Types
+import java.sql.PreparedStatement
+import java.sql.ResultSet
+import java.time.Instant
+import java.time.LocalDate
+import java.time.OffsetDateTime
+import java.time.ZoneId
+import java.util.*
+import java.util.concurrent.TimeUnit
+import java.util.concurrent.locks.ReentrantLock
+import java.util.function.Function
+import java.util.function.Supplier
+
+/**
+ *
+ */
+
+const val MAX_DURATION_MILLIS = 60000L
+const val DATA_DELETION_RATE = 30000L
+
+class BackgroundExpiredDataDeletionService(
+        hazelcastInstance: HazelcastInstance,
+        private val indexerConfiguration: IndexerConfiguration,
+        private val hds: HikariDataSource,
+        private val elasticsearchApi: ConductorElasticsearchApi
+) {
+    companion object {
+        private val logger = LoggerFactory.getLogger(BackgroundExpiredDataDeletionService::class.java)!!
+    }
+
+    private val propertyTypes: IMap<UUID, PropertyType> = hazelcastInstance.getMap(HazelcastMap.PROPERTY_TYPES.name)
+    private val entitySets: IMap<UUID, EntitySet> = hazelcastInstance.getMap(HazelcastMap.ENTITY_SETS.name)
+    private val expirationLocks: IMap<UUID, Long> = hazelcastInstance.getMap(HazelcastMap.EXPIRATION_LOCKS.name)
+
+    //necessary??????
+    init {
+        expirationLocks.addIndex(QueryConstants.THIS_ATTRIBUTE_NAME.value(), true)
+    }
+
+    private val taskLock = ReentrantLock()
+
+    @Suppress("UNCHECKED_CAST", "UNUSED")
+    @Scheduled(fixedRate = MAX_DURATION_MILLIS)
+    fun scavengeIndexingLocks() {
+        expirationLocks.removeAll(
+                Predicates.lessThan(
+                        QueryConstants.THIS_ATTRIBUTE_NAME.value(),
+                        System.currentTimeMillis()
+                ) as Predicate<UUID, Long>
+        )
+    }
+
+    @Suppress("UNUSED")
+    @Scheduled(fixedRate = DATA_DELETION_RATE)
+    fun deleteExpiredDataFromEntitySets() {
+        logger.info("Starting background expired data deletion task.")
+        //Keep number of expired data deletion jobs under control
+        if (taskLock.tryLock()) {
+            try {
+                if (indexerConfiguration.backgroundExpiredDataDeletionEnabled) {
+                    val w = Stopwatch.createStarted()
+                    //We shuffle entity sets to make sure we have a chance to work share and index everything
+                    //lock entity sets we are working on
+                    val lockedEntitySets = entitySets.values
+                            .shuffled()
```

Comment:
I am very unclear on the utility of `.shuffled()` here 😕

Reply:
I lifted it from the backgroundindexingservice, I dunno what it's job is either but I left it to be safe. I can look into it further.

Reply:
Not sure why randomizing the order is necessary, but this will probably be a bit faster if it's moved after the filters (or removed)

Reply:
think I'm just gonna remove it then.

Reply:
actually it may serve a purpose, as the rate at which the task is set to start is half the time allotted to keep the entity sets locked. So I think it might be to prevent getting stuck on the same locked entity sets?